src/EFCore.PG/Metadata/NpgsqlAlterDatabaseOperationAnnotations.cs

```diff
@@ -49,11 +50,131 @@ public NpgsqlAlterDatabaseOperationAnnotations([NotNull] AlterDatabaseOperation
             : base(operation)
             => _oldDatabase = operation.OldDatabase;
 
+        #region Extensions
+
         [NotNull]
         public virtual PostgresExtension GetOrAddPostgresExtension(
             [CanBeNull] string schema,
             [NotNull] string name,
             [CanBeNull] string version)
             => PostgresExtension.GetOrAddPostgresExtension((IMutableAnnotatable)Metadata, schema, name, version);
+
+        [NotNull]
+        public virtual PostgresExtension GetOrAddOldPostgresExtension(
+            [CanBeNull] string schema,
+            [NotNull] string name,
+            [CanBeNull] string version)
+            => PostgresExtension.GetOrAddPostgresExtension((IMutableAnnotatable)_oldDatabase, schema, name, version);
+
+        [NotNull]
+        [ItemNotNull]
+        public virtual IEnumerable<PostgresExtension> GetPostgresExtensionsToCreate()
+            => PostgresExtensions.Where(ne => OldPostgresExtensions.All(oe => oe.Name != ne.Name));
+
+        // TODO: Some forms of extension dropping are actually supported...
+        [NotNull]
+        [ItemNotNull]
+        public virtual IEnumerable<PostgresExtension> GetPostgresExtensionsToDrop()
+            => Enumerable.Empty<PostgresExtension>();
+
+        // TODO: Some forms of extension alterations are actually supported...
+        [NotNull]
+        [ItemNotNull]
+        public virtual IEnumerable<PostgresExtension> GetPostgresExtensionsToAlter()
+            => Enumerable.Empty<PostgresExtension>();
+
+        #endregion
+
+        #region Enums
+
+        [NotNull]
+        public virtual PostgresEnum GetOrAddPostgresEnum(
+            [CanBeNull] string schema,
+            [NotNull] string name,
+            [NotNull] string[] labels)
+            => PostgresEnum.GetOrAddPostgresEnum((IMutableAnnotatable)Metadata, schema, name, labels);
+
+        [NotNull]
+        public virtual PostgresEnum GetOrAddOldPostgresEnum(
+            [CanBeNull] string schema,
+            [NotNull] string name,
+            [NotNull] string[] labels)
+            => PostgresEnum.GetOrAddPostgresEnum((IMutableAnnotatable)_oldDatabase, schema, name, labels);
+
+        [NotNull]
+        [ItemNotNull]
+        public virtual IEnumerable<PostgresEnum> GetPostgresEnumsToCreate()
+            => PostgresEnums.Where(ne => OldPostgresEnums.All(oe => oe.Name != ne.Name));
```

Comment:
After trying a couple of annotation-based implementations, I decided to take a conservative approach for now, in which we just move the existing logic from `NpgsqlMigrationsSqlGenerator` into a more appropriate location here.

If we later decide to store created/dropped types as a special form of annotation (e.g. `Npgsql:DroppedType:....`), we would just update these methods to do the lookup/return rather than doing the diff directly.